corgidrp/l4_to_tda.py

```diff
@@ -137,6 +137,516 @@ def determine_color_cor(input_dataset, ref_star, source_star):
+    host_wave = host_cal[0].hdu_list['SPEC_WAVE'].data
+    comp_wave = comp_cal[0].hdu_list['SPEC_WAVE'].data
+
+    # Align wavelength grids if the host/companion spectra were sampled in opposite directions
+    # (np.interp requires increasing x-coordinates).
+    if not np.allclose(host_wave, comp_wave):
+        host_decreasing = host_wave[0] > host_wave[-1]
+        comp_decreasing = comp_wave[0] > comp_wave[-1]
+
+        host_wave_interp = host_wave[::-1] if host_decreasing else host_wave
+        if comp_decreasing:
+            comp_wave_interp = comp_wave[::-1]
+            comp_spec_interp = comp_spec[::-1]
+            comp_err_interp = comp_err[::-1]
+        else:
+            comp_wave_interp = comp_wave
+            comp_spec_interp = comp_spec
+            comp_err_interp = comp_err
+
+        comp_spec_aligned = np.interp(
+            host_wave_interp,
+            comp_wave_interp,
+            comp_spec_interp,
+            left=comp_spec_interp[0],
+            right=comp_spec_interp[-1]
+        )
+        comp_err_aligned = np.interp(
+            host_wave_interp,
+            comp_wave_interp,
+            comp_err_interp,
+            left=comp_err_interp[0],
+            right=comp_err_interp[-1]
+        )
+        comp_spec = comp_spec_aligned[::-1] if host_decreasing else comp_spec_aligned
+        comp_err = comp_err_aligned[::-1] if host_decreasing else comp_err_aligned
+
+    # Ratio calculation - invalid host/companion values become NaN rather than raising warnings.
+    with np.errstate(divide='ignore', invalid='ignore'):
+        flux_ratio = comp_spec / host_spec
+        invalid_mask = (host_spec == 0) | ~np.isfinite(host_spec) | ~np.isfinite(comp_spec)
+        flux_ratio[invalid_mask] = np.nan
+
+    # Propagate uncertainties for comp/host division:
+    # σ_R^2 = (σ_C / H)^2 + (C * σ_H / H^2)^2
+    ratio_unc = np.full_like(flux_ratio, np.nan, dtype=float)
+    valid_unc = (
+        (host_spec != 0) &
+        np.isfinite(host_spec) &
+        np.isfinite(comp_spec) &
+        np.isfinite(host_err) &
+        np.isfinite(comp_err)
+    )
+    if np.any(valid_unc):
+        comp_term = np.zeros_like(flux_ratio, dtype=float)
+        host_term = np.zeros_like(flux_ratio, dtype=float)
+        comp_term[valid_unc] = (comp_err[valid_unc] / host_spec[valid_unc]) ** 2
+        host_term[valid_unc] = (
+            (comp_spec[valid_unc] * host_err[valid_unc]) / (host_spec[valid_unc] ** 2)
+        ) ** 2
+        variance = comp_term + host_term
+        ratio_unc[valid_unc] = np.sqrt(variance[valid_unc])
+
+    metadata = {
+        'roll': host_image.pri_hdr.get('ROLL'),
+        'companion_roll': companion_image.pri_hdr.get('ROLL'),
+        'exp_time': host_image.pri_hdr.get('EXP_TIME', 1.0),
+        'weighted': False,
+        'ratio_err': ratio_unc,
+    }
+
+    return flux_ratio, host_wave, metadata
+
+
+def compute_weighted_spec_flux_ratio(host_datasets, companion_datasets, fluxcal_factor,
```

Comment:
I think I just meant a function that given a single input Dataset, combines them together into a single spectrum (weighted based on errors, and interpolated to a common wavelength grid). The idea is that we could use this to separately run on any dataset of spectrum (Whether host star or companion, whether in flux ratio units or not), independent of the other functions above. 

My previous concern with doing everything in a single function is that I'm not sure we are guaranteed to get one host star spectrum per roll. So there may not be this 1:1 mapping.